hpct-ldap-server/src/charm.py

```diff
@@ -0,0 +1,136 @@
+        self.framework.observe(self.on.add_user_action, self._on_add_user_action)
+        self.framework.observe(self.on.set_config_action, self._on_set_config_action)
+        # Integrations
+        self.framework.observe(
+            self.on.tls_cert_relation_changed, self._on_tls_cert_relation_changed
+        )
+        # Server Manager
+        self.ldapserver_manager = LdapServer()
+
+    def _on_add_group_action(self, event):
+        """Handle add-group action."""
+        if not self.unit.is_leader():
+            event.fail("The action can be run only on leader unit.")
+            return
+        replicas = self.model.get_relation("replicas")
+        domain = replicas.data[self.app].get("domain")
+        self.ldapserver_manager.add_group(
+            event.params["admin-passwd"],
+            domain,
+            event.params["gid"],
+            event.params["group"],
+        )
+
+    def _on_add_user_action(self, event):
+        """Handle add-user action."""
+        if not self.unit.is_leader():
+            event.fail("The action can be run only on leader unit.")
+            return
+        replicas = self.model.get_relation("replicas")
+        domain = replicas.data[self.app].get("domain")
+        self.ldapserver_manager.add_user(
+            event.params["admin-passwd"],
+            domain,
+            event.params["gecos"],
+            event.params["gid"],
+            event.params["homedir"],
+            event.params["login"],
+            event.params["passwd"],
+            event.params["uid"],
+            event.params["user"],
+        )
+
+    def _on_install(self, event):
+        """Handle install event."""
+        logger.info("Install")
+        self.ldapserver_manager.install()
+
+    def _on_set_config_action(self, event):
+        """Handle set-config action."""
+        if not self.unit.is_leader():
+            event.fail("The action can be run only on leader unit.")
+            return
+        self.ldapserver_manager.set_config(
+            event.params["admin-passwd"],
+            event.params["domain"],
+            event.params["org"],
+        )
+        self.ldapserver_manager.tls_gen(event.params["org"])
+        replicas = self.model.get_relation("replicas")
+        replicas.data[self.app].update(
+            {
+                "domain": event.params["domain"],
+                "org": event.params["org"],
+            }
+        )
+
+    def _on_start(self, event):
+        """Handle start event."""
+        self.ldapserver_manager.start()
+        self.unit.status = ActiveStatus("LDAP Server Started")
+
+    def _on_tls_cert_relation_changed(self, event):
+        """Handle the tls-transfer action."""
+        # Get TLS relation
+        tls_relation = self.model.get_relation("tls-cert")
+        # Get Replicas relation
+        replicas = self.model.get_relation("replicas")
+        domain = replicas.data[self.app].get("domain")
+        if not tls_relation:
```

Comment:
Will this ever happen if the relation is not created?

Reply:
No, removing if statement